src/apps/MASSVEC3DPA-Cuda.cpp

```diff
@@ -0,0 +1,855 @@
+                           block_size) {
+      MASSVEC3DPA_4;
+    }
+    __syncthreads();
+
+    GPU_SHARED_LOOP_3D_INC(qx, qy, qz, MVPA_Q1D, MVPA_Q1D, MVPA_Q1D,
+                           block_size) {
+      MASSVEC3DPA_5;
+    }
+    __syncthreads();
+
+    GPU_SHARED_LOOP_3D_INC(dx, qy, qz, MVPA_D1D, MVPA_Q1D, MVPA_Q1D,
+                           block_size) {
+      MASSVEC3DPA_6;
+    }
+    __syncthreads();
+
+    GPU_SHARED_LOOP_3D_INC(dx, dy, qz, MVPA_D1D, MVPA_D1D, MVPA_Q1D,
+                           block_size) {
+      MASSVEC3DPA_7;
+    }
+    __syncthreads();
+
+    GPU_SHARED_LOOP_3D_INC(dx, dy, dz, MVPA_D1D, MVPA_D1D, MVPA_D1D,
+                           block_size) {
+      MASSVEC3DPA_8;
+    }
+    __syncthreads();
+
+  } // (c) dimension loop
+}
+
+template <size_t block_size>
+__launch_bounds__(block_size) __global__
+    void MassVec3DPA_DIRECT(const Real_ptr B, const Real_ptr Bt,
+                            const Real_ptr D, const Real_ptr X, Real_ptr Y) {
+
+  const Index_type e = blockIdx.x;
+
+  MASSVEC3DPA_0_GPU;
+
+  GPU_SHARED_DIRECT_2D(q, d, MVPA_Q1D, MVPA_D1D) { MASSVEC3DPA_1; }
+
+  for (Index_type c = 0; c < 3; ++c) {
+    GPU_SHARED_DIRECT_3D(dx, dy, dz, MVPA_D1D, MVPA_D1D, MVPA_D1D) {
+      MASSVEC3DPA_2;
+    }
+    __syncthreads();
+
+    GPU_SHARED_DIRECT_3D(qx, dy, dz, MVPA_Q1D, MVPA_D1D, MVPA_D1D) {
+      MASSVEC3DPA_3;
+    }
+    __syncthreads();
+
+    GPU_SHARED_DIRECT_3D(qx, qy, dz, MVPA_Q1D, MVPA_Q1D, MVPA_D1D) {
+      MASSVEC3DPA_4;
+    }
+    __syncthreads();
+
+    GPU_SHARED_DIRECT_3D(qx, qy, qz, MVPA_Q1D, MVPA_Q1D, MVPA_Q1D) {
+      MASSVEC3DPA_5;
+    }
+    __syncthreads();
+
+    GPU_SHARED_DIRECT_3D(dx, qy, qz, MVPA_D1D, MVPA_Q1D, MVPA_Q1D) {
+      MASSVEC3DPA_6;
+    }
+    __syncthreads();
+
+    GPU_SHARED_DIRECT_3D(dx, dy, qz, MVPA_D1D, MVPA_D1D, MVPA_Q1D) {
+      MASSVEC3DPA_7;
+    }
+    __syncthreads();
+
+    GPU_SHARED_DIRECT_3D(dx, dy, dz, MVPA_D1D, MVPA_D1D, MVPA_D1D) {
+      MASSVEC3DPA_8;
+    }
+    __syncthreads();
+
+  } // (c) dimension loop
+}
+
+template <size_t block_size>
+void MASSVEC3DPA::runCudaVariantImpl(VariantID vid, size_t tune_idx) {
+  const Index_type run_reps = getRunReps();
+
+  auto res{getCudaResource()};
+
+  MASSVEC3DPA_DATA_SETUP;
+
+  switch (vid) {
+
+  case Base_CUDA: {
+
+    if (tune_idx == 0) {
+
+      startTimer();
+      for (RepIndex_type irep = 0; irep < run_reps; irep = irep + 1) {
+
+        dim3 nthreads_per_block(MVPA_Q1D, MVPA_Q1D, MVPA_Q1D);
+        constexpr size_t shmem = 0;
+
+        RPlaunchCudaKernel((MassVec3DPA_BLOCKDIM_LOOP_INC<block_size>), NE,
+                          nthreads_per_block, shmem, res.get_stream(), B, Bt, D,
+                          X, Y);
+      }
+      stopTimer();
+
+      // Loop constants
+    } else if (tune_idx == 1) {
+
+      // Mark volatile because we want the value to be treated as a runtime
+      // value
+      volatile Index_type runtime_loop_bounds = MVPA_Q1D;
```

Comment:
Volatile is not they way to do this. It will force the value to be read in from memory (not cached) each time it is used. How about adding a member variable to the class that is set to this in the constructor?

Reply:
fixed